### PR TITLE
zpi-77 - flattening of the role graph

### PR DESCRIPTION
**Pull request by Nightbaka**
Forget about graphs, its time for matrices

### CODE REVIEW

**Comment by Nightbaka** [comment]
I've changed my mind, I will resubmit this PR changes when I have the entire endpoint ready, to not make you go through this multiple times

**Review by samuelus — CHANGES_REQUESTED** [review]
I'll do a review later, but I want to have dibs so it doesn't go without my approval. Btw please don't merge it with this name XD

File: backend/auth/role_map_repository_test.go [hunk 1]
```diff
@@ -223,3 +239,541 @@ func TestHasPermission(t *testing.T) {
...
+					log.Printf("  Subroles: %v", subrole.value.Subroles)
+					log.Printf("  Permit: %v", subrole.value.Permit)
+					log.Printf("  Deny: %v", subrole.value.Deny)
+				}
+			}
+		})
+	}
+}
+
+func createFuzzyRoleRepo(permits, denies, resources, namespaces int, pstar, dstar float64, operations []models.OperationType, graph map[string][]string) *RoleMapRepository {
```
Comment (Ninja00Shadow) [inline backend/auth/role_map_repository_test.go:694]: What is this? To make tests easier?

File: backend/auth/role_map_repository.go [hunk 2]
```diff
@@ -15,7 +16,8 @@ type RoleMapRepository struct {
 	// Subrole map is optional, if not provided, subroles will be ignored. Roles received in token are checked only with RoleMap
 	RoleMap map[string]*models.Role
 	SubroleMap map[string]*models.Role
-	flattenedMap map[string]map[string]map[string][]models.OperationType
+	flattenedMap map[string]map[string]map[string]map[models.OperationType]struct{}
```
Comment (Ninja00Shadow) [inline backend/auth/role_map_repository.go:19]: This is a nightmare fuel.

**Review by Ninja00Shadow — COMMENTED** [review]
This is too much for me.

File: backend/auth/role_map_repository_test.go (hunk 1 above)
Comment (Nightbaka) [reply backend/auth/role_map_repository_test.go:694]: This is generating a random role map, with given parameters. Its used for fuzzy tests - tests with random inputs. I decided this approach would be better than hardcoded test cases to avoid missing out any edge cases

File: backend/auth/role_map_repository.go (hunk 2 above)
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:19]: you can think of it as a map of role -> permission matrix

Comment (Nightbaka) [reply backend/auth/role_map_repository.go:19]: I had no nightmares about that while coding so can't confirm this is a valid comment

File: backend/auth/role_map_repository.go [hunk 3]
```diff
@@ -58,99 +59,227 @@ func getOrDefaultEnv(key, defaultValue string) string {
...
+			for _, operations := range matrix[deny.Namespace] {
+				removeOp(operations, deny.Type)
+			}
+		} else { // n r op
+			removeOp(matrix[deny.Namespace][deny.Resource], deny.Type)
 		}
 	}
-
-	return result
 }
```
[the remark this answers is not in the document]
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:239]: Yes but I would rather keep these 2 separate. I could refactor this into function accepting func operating on operations, but I would rather have them like this

[the remark this answers is not in the document]
Comment (Ninja00Shadow) [reply backend/auth/role_map_repository.go:239]: Can you at least separate this to another function? This should be easy:

```
_, hasNamespace := matrix[permit.Namespace]
	_, hasResource := matrix["*"][permit.Resource]

	if !hasNamespace {
		expandNamespaces(permit.Namespace, matrix)
	}
	if !hasResource {
		expandResources(permit.Resource, matrix)
	}
```

File: backend/auth/role_map_repository_test.go (hunk 1 above)
Comment (Ninja00Shadow) [reply backend/auth/role_map_repository_test.go:694]: Understandable, have a great day

File: backend/auth/role_map_repository.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:239]: I would prefer not to do that, I think it would actually worsen, the readability. I will add 6 line shorter version instead. Will that suffice?

**Review by Ninja00Shadow — APPROVED** [review]

**Comment by Ninja00Shadow** [comment]
image

File: backend/auth/role_map_repository.go (hunk 2 above)
Comment (samuelus) [reply backend/auth/role_map_repository.go:19]: It looks like a nightmare, but under the hood it's brilliant. I really admire you for finding this optimization

File: backend/auth/role_map_repository_test.go
Comment (samuelus) [inline backend/auth/role_map_repository_test.go:1]: They have high coverage so i will assume they are good xd

Comment (samuelus) [reply backend/auth/role_map_repository_test.go:1]: But you can add this case to TestHasCycle, to get full coverage of this function. In the future for coverage raport you can use this commands:
```
go test -coverprofile cover.out
go tool cover -html=cover.out
```

```
,
		{
			name: "Non-existing subrole",
			roleMap: map[string]*models.Role{
				"admin": {Subroles: []string{"nonexistent"}},
			},
			expected: false,
		},
```

**Review by samuelus — CHANGES_REQUESTED** [review]
I know I gave a lot of comments, but I'm really impressed how good this code is, looking at its complexity

File: backend/auth/role_map_repository.go [hunk 4]
```diff
@@ -33,18 +35,17 @@ func GetRoleMapInstance() (*RoleMapRepository, error) {
 			log.Printf("Failed to initialize RoleMapRepository")
```
[the remark this answers is not in the document]
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:35]: Sure, there are more detailed errors logged in the GetRoleMapConfig anyways.

File: backend/auth/role_map_repository.go [hunk 5]
```diff
@@ -58,99 +59,221 @@ func getOrDefaultEnv(key, defaultValue string) string {
...
-		make(map[string]struct{}),
-	)
+func addMatrix(m1 map[string]map[string]map[models.OperationType]struct{}, 
+	m2 map[string]map[string]map[models.OperationType]struct{}) (
+		map[string]map[string]map[models.OperationType]struct{}) {
+	x1, y1 := len(m1), len(m1["*"])
+	x2, y2 := len(m2), len(m2["*"])
+	if x1 * y1 < x2 * y2 {
+		m1, m2 = m2, m1
+	}
```
[the remark this answers is not in the document]
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:134]: There is no point besides readability and writability. This is too clutered with syntax for my taste:
`if len(m1) * len(m1["*"]) < len(m2) * len(m2["*"])  {
    m1, m2 = m2, m1
}`

File: backend/auth/role_map_repository.go [hunk 6]
```diff
@@ -58,99 +59,221 @@ func getOrDefaultEnv(key, defaultValue string) string {
...
+			for _, operations := range matrix[deny.Namespace] {
+				removeOp(operations, deny.Type)
+			}
+		} else { // n r op
+			removeOp(matrix[deny.Namespace][deny.Resource], deny.Type)
 		}
 	}
-
-	return result
 }
```
[the remark this answers is not in the document]
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:233]: Ehhh it could be I just still can't believe the same exact logic for both of these works

File: backend/auth/role_map_repository.go [hunk 7]
```diff
@@ -58,99 +59,221 @@ func getOrDefaultEnv(key, defaultValue string) string {
...
+	for _, namespaced := range matrix {
+		namespaced[resource] = make(map[models.OperationType]struct{})
+		for opType := range namespaced["*"] {
+			namespaced[resource][opType] = struct{}{}
+		}
+	}
 }
-	
 
 func hasPermission(role *models.Role, subroleMap map[string]*models.Role, operation *models.Operation, visited map[string]struct{}) bool {
```
[the remark this answers is not in the document]
Comment (Nightbaka) [reply backend/auth/role_map_repository.go:278]: I need it for comparison tests atleast. Was thinking of using it as fallback if initializing the matrix would take a longer time then expected to, but honestly propably won't do that. Something about moving it to the test file doesn't sit right in me, but if you insist it could be done

File: backend/auth/role_map_repository.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ninja00Shadow) [reply backend/auth/role_map_repository.go]: > Yes but I would rather keep these 2 separate. I could refactor this into function accepting func operating on operations, but I would rather have them like this

Traitor.

[the remark this answers is not in the document]
Comment (Nightbaka) [reply backend/auth/role_map_repository.go]: Samuel gave me a ready refactor,  you just yyyyyy... can it be simplified?

[the remark this answers is not in the document]
Comment (Nightbaka) [reply backend/auth/role_map_repository.go]: Yes I was lazy

File: backend/auth/role_map_repository.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (samuelus) [reply backend/auth/role_map_repository.go:246]: If it is a function used only in tests, then it should be in the test file, because now it looks like it is not used anywhere (for tests I just thought it is used in its tests, not in their creation), plus it spoils the test coverage, as if someone would ever look at it more

**Review by samuelus — APPROVED** [review]
